scipy/special/_special_ufuncs.cpp

```diff
@@ -206,9 +206,15 @@ extern const char *yve_doc;
 extern "C" int wrap_PyUFunc_getfperr() { return PyUFunc_getfperr(); }
 
 static PyModuleDef _special_ufuncs_def = {
-    .m_base = PyModuleDef_HEAD_INIT,
-    .m_name = "_special_ufuncs",
-    .m_size = -1,
```

Comment:
It's a pity, I like the designated initializer syntax much better. All supported (non-EOL) MSVC versions have full support for C++20, so I'd be open to consider using C++20 on windows (without touching the C++17 baseline for now, yet still enabling designated initializer use because GCC/clang have it as an extension in earlier standard versions).

Reply:
I'd be 100% onboard with changing the standard to C++20 on Windows. Can we do this? Does it need a discussion first?

Reply:
As long as @rgommers doesn't object, I'd consider this settled. I spent quite a bit of time on the various standards and features supported across the compiler landscape, and using C++20 has no cost in terms of compatibility.

The only "price" is would be using different standards between windows and unix.

Reply:
Different standard across platforms doesn't sound like a healthy idea to me, nor is it technically correct to then use designated initializers - we don't only worry about what we have in CI, but also niche compilers that we can't test. And our roadmap says _"Given how big of a release the C++20 standard was, it is expected that it will take a [while yet](https://discourse.llvm.org/t/rfc-clang-17-0-6-would-be-minimum-version-to-build-llvm-in-c-20/75345/8) before we can start considering moving our baseline."_

This is a quite trivial change and we have many `PyModuleDef`'s that look like this across the code base. So nothing much is lost by sticking with C++17.

Last reason that comes to mind: we're going to get other platform-specific differences - I've already seen one recently with Pythran. The bump to C17 also wasn't smooth, there's always some hiccup.

Reply:
Yeah, I wrote that line. :)

But the roadmap talks about C++20 support across all compilers - the situation can be different for a single one.

> but also niche compilers that we can't test.

That's an interesting point - we do make quite a bit of effort to survey the platform landscape, but we say nothing about niche compilers. Is there any platform we care about that doesn't use gcc/clang/msvc?

> Different standard across platforms doesn't sound like a healthy idea to me

I agree of course, though in this case I'm tempted to think the benefits might outweigh the costs. Initialising structs without designated initializers is a minefield, whereas with them the intent becomes perfectly obvious & safe. I prefer the added safety, and would be willing to accept the cost of shaking out some C++20 bugs.

Reply:
> Is there any platform we care about that doesn't use gcc/clang/msvc?

Sure:
- Intel compilers are a major one we don't currently test (Spack does have CI for SciPy + oneAPI),
- Emscripten comes to mind (for Pyodide support), we don't test it in CI yet, but hopefully will in the near future.
- AIX uses both GCC and IBM's compilers (XL C/C++). The latter we've regularly broken in the past, not sure what the status is now.
- I'm aware that there are proprietary ones too, like Fujitsu has one for the Fukagu supercomputer - no idea how different it is, just that at least in NumPy there's been some PRs to support it in the past.

Reply:
Alright, consider the suggestion withdrawn, seems we'll wait a couple more years for designated initializers 🥲

BTW, the XL compilers are actually mentioned in the toolchain, but we could add Intel and Emscripten. I'll add it to https://github.com/scipy/scipy/issues/19815